pypeit/core/flexure.py

```diff
@@ -303,147 +305,226 @@ def spec_flex_shift(obj_skyspec, arx_skyspec, arx_lines, mxshft=20):
+    nslits = slits.nslits
+    gpm = np.logical_not(bpm)
+    gdslits = np.where(gpm)[0]
 
-    # Loop on objects
+    # Initialise the flexure list for each slit
     flex_list = []
-
     # Slit/objects to come back to
     return_later_sobjs = []
 
-    # Loop over slits, and then over objects here
+    # Loop over slits, and then over objects
     for islit in range(nslits):
-        i_slitord = slitord[islit]
-        msgs.info("Working on flexure in slit (if an object was detected): {:d}".format(islit))
+        msgs.info("Working on spectral flexure of slit: {:d}".format(islit))
 
-        indx = specobjs.slitorder_indices(i_slitord)
-        this_specobjs = specobjs[indx]
         # Reset
         flex_dict = dict(polyfit=[], shift=[], subpix=[], corr=[],
                          corr_cen=[], spec_file=sky_file, smooth=[],
                          arx_spec=[], sky_spec=[])
+
         # If no objects on this slit append an empty dictionary
         if islit not in gdslits:
             flex_list.append(flex_dict.copy())
             continue
-        for ss, specobj in enumerate(this_specobjs):
-            if specobj is None:
-                continue
-            if specobj['BOX_WAVE'] is None: #len(specobj._data.keys()) == 1:  # Nothing extracted; only the trace exists
-                continue
-            msgs.info("Working on flexure for object # {:d}".format(specobj.OBJID) + "in slit # {:d}".format(islit))
-            # Using boxcar
-            if method in ['boxcar', 'slitcen']:
-                sky_wave = specobj.BOX_WAVE #.to('AA').value
-                sky_flux = specobj.BOX_COUNTS_SKY
-            else:
-                msgs.error("Not ready for this flexure method: {}".format(method))
 
-            # Generate 1D spectrum for object
-            obj_sky = xspectrum1d.XSpectrum1D.from_tuple((sky_wave, sky_flux))
+        if slit_cen:
+            slit_spat = slits.spat_id[islit]
+            sky_wave = slitspecs[islit].wavelength.value
+            sky_flux = slitspecs[islit].flux.value
 
             # Calculate the shift
-            fdict = spec_flex_shift(obj_sky, sky_spectrum, sky_lines, mxshft=mxshft)
-            punt = False
-            if fdict is None:
-                msgs.warn("Flexure shift calculation failed for this spectrum.")
-                if sv_fdict is not None:
-                    msgs.warn("Will used saved estimate from a previous slit/object")
-                    fdict = copy.deepcopy(sv_fdict)
+            fdict = spec_flex_shift(slitspecs[islit], sky_spectrum, sky_lines, mxshft=mxshft)
+            # Failed?
+            if fdict is not None:
+                # Update dict
+                for key in ['polyfit', 'shift', 'subpix', 'corr', 'corr_cen', 'smooth', 'arx_spec']:
+                    flex_dict[key].append(fdict[key])
+                # Interpolate
+                sky_wave_new = specobj.SpecObj.flexure_interp(fdict['shift'], sky_wave)
+                flex_dict['sky_spec'].append(
```

Comment:
If anything this should be operating on specobjs objects and not a line tools object. It seems silly to me that we are calling a method in specobj and then appending stuff to an Xspectrum1d. Only one object is needed here.